crates/common/consensus/lean/src/state.rs

```diff
@@ -1,14 +1,90 @@
+use alloy_primitives::B256;
 use serde::{Deserialize, Serialize};
 use ssz_derive::{Decode, Encode};
-use ssz_types::VariableList;
+use ssz_types::{
+    BitList, VariableList,
+    typenum::{U4096, U262144, U16777216, Unsigned},
+};
 use tree_hash_derive::TreeHash;
 
-use crate::validator::Validator;
+use crate::config::Config;
 
-#[derive(Debug, PartialEq, Clone, Serialize, Deserialize, Encode, Decode, TreeHash)]
-pub struct BeamState {
-    pub genesis_time: u64,
+#[derive(Debug, Eq, PartialEq, Clone, Serialize, Deserialize, Encode, Decode, TreeHash)]
+pub struct LeanState {
+    pub config: Config,
 
-    /// Up to 1 million validators
-    pub validators: VariableList<Validator, ssz_types::typenum::U1000000>,
+    pub latest_justified_hash: B256,
+    pub latest_justified_slot: u64,
+    pub latest_finalized_hash: B256,
+    pub latest_finalized_slot: u64,
+
+    pub historical_block_hashes: VariableList<B256, U262144>,
+    pub justified_slots: VariableList<bool, U262144>,
+
+    // Diverged from Python implementation:
+    // Originally `justifications: Dict[str, List[bool]]`
+    pub justifications_roots: VariableList<B256, U262144>,
+    pub justifications_roots_validators: BitList<U16777216>,
+}
+
+impl LeanState {
+    fn get_justifications_roots_index(&self, root: &B256) -> Option<usize> {
+        self.justifications_roots.iter().position(|r| r == root)
+    }
+
+    fn get_justifications_roots_range(&self, index: &usize) -> (usize, usize) {
+        // Start from index * MAX_HISTORICAL_BLOCK_HASHES, ends at start + VALIDATOR_REGISTRY_LIMIT
+        let start_range = index * U262144::to_usize();
+        let end_range = start_range + U4096::to_usize();
```

Comment:
```suggestion
        let start_range = index * 262144;
        let end_range = start_range + 4096;
```

Can't we just do this?